src/Http2Client.ts

```diff
@@ -40,38 +44,82 @@ export class Http2Client extends EventEmitter implements Client {
 	 *
 	 * @param host - hostname or IP address of daemon
 	 * @param port - port number of daemon on the host
-	 * @override
 	 */
-	introduce(host: string, port: number): void {
+	public introduce(host: string, port: number): void {
 		const client = connect(`http://${hostAndPort(host, port)}`)
 		client.on('error', (err) => this.emit('error', err))
-		this.available(client)
+		this.setAvailableAndCheckJobs(client)
 	}
 
 	/**
-	 * Drive the given daemon, by providing it with one task at a time in an async loop.
+	 * Mark the daemon as available.
+	 * Then calls function to check if there are any doable jobs.
 	 *
-	 * @param client - a HTTP/2 client connected to a daemon.
+	 * @param daemon - The daemon that is available.
+	 */
+	private setAvailableAndCheckJobs(daemon: ClientHttp2Session): void {
+		this.availableDaemons.add(daemon)
+		this.checkJobsAndAssign()
+	}
+
+	/**
+	 * Tries to assign the next available job to a daemon.
+	 * Closes all daemons if all jobs are finished.
 	 */
-	private available(client: ClientHttp2Session): void {
-		const job = this.queue.pop()
-		if (!job) {
-			client.close(() => this.emit('done'))
-			return
+	private checkJobsAndAssign(): void {
+		const job = this.jobOrderer.popNextJob()
+
+		if (job && this.availableDaemons.size > 0) {
+			const daemon: ClientHttp2Session = this.availableDaemons.values().next()
```

Comment:
Take a randomly chosen daemon, not the next available. That's an easy way to ensure we distribute jobs evenly among the daemons.

Reply:
Getting a random element from a `Set` is O(n). I could have an array or something that also contains the daemons for random access.

What do you mean by evenly? Do you think that the daemons will come to us in a significantly nonrandom order?

Reply:
> Getting a random element from a `Set` is O(n).

Good point, I'm not sure atm how to cleanly solve this. (since this isn't performance critical, it's ok to do the O(n) operation for now with a comment stating the future optimization).

> Do you think that the daemons will come to us in a significantly nonrandom order?

The daemons may come to us in any order, and that order probably isn't random. Besides, even if it is, the issue is that the order is stable (unless the Set ordering is internally randomized like with Python's dict, in which case this is less important).

So as a long-running DAG is being worked on, machine 1 would get 100% utilization, machine 2 only when machine 1 is busy so maybe 50% depending on the amount of parallelism in the graph, and so on with the last machine getting no use (for example).

Reply:
Ah I just found the [MDN docs for Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set):
> You can iterate through the elements of a set in insertion order.

That's interesting. I guess that means that we would pop the next available daemon from the front, then return it to the back when we're done. You might want to confirm this is really what happens when you remove() and re-add(). If so, perhaps that gives us enough of an even-distribution guarantee?

Reply:
I don't think that we should depend on `Set`'s implementation as that can change at any time.

Also, if you want to equally distribute work, it could be better to actually keep track of how many jobs each Daemon has and assign using that. Delegating to randomness seems kinda inefficient.

Regardless, I think that this should be part of another PR, since the focus of this PR isn't even the Client.

Reply:
> I don't think that we should depend on `Set`'s implementation as that can change at any time.

It's not an implementation detail of Set, it appears to be part of its specified behavior, so it's fine to rely on it. No changes are necessary though, the current code most likely does what we want (pops from the "front" of the Set in insertion order, pushes to the "back").

> Delegating to randomness seems kinda inefficient.

It would actually be more efficient - you end up distributing evenly without having to maintain any data structures. But as explained above it's not needed here because of the way Set works.